realtek: fix ethtool stats for RTL839x and RTL930x

janh opened the pull request:
The MIB registers contain different stats depending on the SoC, and for RTL930x some stats are in an additional register.

RTL931x remains unsupported, because it uses a table and thus requires a separate implementation.

Tested on: HPE 1920-8G (rtl838x), HPE-1920-48G (rtl839x), Zyxel XGS1010-12 (rtl930x)
The reported stats seem plausible on all devices.

janh committed:
realtek: fix ethtool stats for RTL839x and RTL930x

The MIB registers contain different stats depending on the SoC, and for
RTL930x some stats are in an additional register.

Create separate MIB descs for each SoC to implement this. Also make
reading 64-bit counters more robust, by protecting against an overflow
of the lower 32 bits during the read.

RTL931x remains unsupported, because it uses a table and thus requires
a separate implementation.

While we are at it, rename structs/functions to use the rtldsa prefix.

Signed-off-by: Jan Hoffmann <jan@3e8.eu>
Link: https://github.com/openwrt/openwrt/pull/18415
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

plappermaul reviewed (approved):
Quick test on LGS310C (RTL838X) worked fine and showed reasonable counters - at least for tx/rx bytes/packets and 1024to1518Octets. So approved from my side.

If you have time and want to make it 100% perfect you could align with naming of the other DSA drivers. E.g. they use Dot3StatsSingleCollisionFrames while we use .3SingleCollisionFrames.

plappermaul commented:
@janh can you assist in testing https://github.com/openwrt/openwrt/pull/18397 and https://github.com/openwrt/openwrt/pull/18402 ? Thanks in advance.

Ansuel commented:
What is the meaning of the .3 values??? Also it would be ideal if we stop custom implementing the MIB that way and move on to more standard way with rmon and other standard mib.

Neustradamus commented:
To follow this PR :)

janh committed:
realtek: use more specific APIs for ethtool stats where possible

The kernel offers several alternatives to get_ethtool_stats which allow
to report some stats in a more structured way. Use them where possible.

Ideally, we should follow RFC 3635 to translate the hardware counters to
the supported frame and octet counters. However, this is not feasible,
as some of the counters are 32-bit only (so it would produce incorrect
results as soon as one of them overflows).

Signed-off-by: Jan Hoffmann <jan@3e8.eu>
Link: https://github.com/openwrt/openwrt/pull/18415
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

janh committed:
realtek: implement get_stats64

By default, the network interface stats are based on software counters,
which only consider traffic from and to the CPU. Implementing the
get_stats64 method allows to report the full hardware counters instead.

Signed-off-by: Jan Hoffmann <jan@3e8.eu>
Link: https://github.com/openwrt/openwrt/pull/18415
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

plappermaul commented:
A look at chipdef.c or chip.c in the GPL source drops gives platform specific maximum settings in rtlXXXX_capacityInfo structures. I guess you need:

- RTL838x: max_frame_len = 10240
- RTL839x: max_frame_len = 12288
- RTL930x: max_frame_len = 12288
- RTL931x: max_frame_len = 12288

E.g. these are checked via macro HAL_MAX_ACCEPT_FRAME_LEN() in dal_<platform>_stat_flexibleCntRange_set() that can be used to setup a flexible MIB counter for individual frame sizes.

Ansuel commented:
Normally the max is based on the MTU value. I would expect a switch to drop anything beyond that.

janh commented:
@plappermaul: I did some testing and can confirm these values for RTL839x and RTL930x. But for RTL838x, only packets up to 10000 bytes are counted.

The code is now updated to use different rmon ranges depending on the platform. For RTL930x, I have added the "PktsOverMaxOctets" counters as an additional range.

I also changed the total frame/packet counters to avoid double counting oversize (over MTU) packets.

janh commented:
Thinking a bit more about this code, I noticed that the calculation of stats based on multiple hardware counters is problematic: As soon as one of the 32-bit counters has an overflow, the result is broken.

I think the only way to solve that completely would be to maintain 64 bit counters in the driver by polling and cumulating the change (just need to poll often enough to catch every 32 bit overflow). I'll try that out to see if it is feasible.

Ansuel commented:
Honestly overflow with counter is a known problem. But normally the 32bit
counter are for error and diag. The packet and bytes are 64bit and is
enough. The alternative is if the thing support interrupt the switch have
always an option to fire interrupt when an overflow occurr or when the
counter are half full... That + local counter is a way to handle the
problem. I feel poll is a bit expensive and not worth to handle.

Il Lun 28 Apr 2025, 18:44 Jan Hoffmann ***@***.***> ha
scritto:

> *janh* left a comment (openwrt/openwrt#18415)
> <https://github.com/openwrt/openwrt/pull/18415#issuecomment-2835859933>
>
> Thinking a bit more about this code, I noticed that the calculation of
> stats based on multiple hardware counters is problematic: As soon as one of
> the 32-bit counters has an overflow, the result is broken.
>
> I think the only way to solve that completely would be to maintain 64 bit
> counters in the driver by polling and cumulating the change (just need to
> poll often enough to catch every 32 bit overflow). I'll try that out to see
> if it is feasible.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/18415#issuecomment-2835859933>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQSNPPHQNZ6MBGYOFX323ZLIFAVCNFSM6AAAAAB2OUFO7CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMZVHA2TSOJTGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

plappermaul commented:
And even if these counters overflow. No need to provide more data than the hardware can. External monitoring tools will handle overflows generously.

janh commented:
A normal counter overflow would obviously not be a problem.

But the current code for the packet counters is actually adding values of multiple hardware counters, some of which are 64 bits and some are 32 bits (depending on the platform), and reports the result as a 64 bits value.

So the reported counters wouldn't actually overflow (which is pretty much impossible for a 64 bit counter anyway), but it would jump around weirdly. The only way that a monitoring tool could detect this would be by ignoring implausible-looking values.

The byte counters are even worse: Here a compensation depending on the packet counter is applied (following RFC3635, inspired by the upstream rtl8365mb driver). But after thinking about this, I really don't think this is a sane thing to do considering the packet counters are unreliable due to overflows.

So my plan would be to either simplify the counters (use only the three main unicast, multicast, broadcast counters for the total packet counters, and remove the adjustment of the byte counters), or implement the polling to get correct 64 bits counters on all platforms.

I looked around in existing upstream drivers while implementing this, and there seem to be several which already use polling to read hardware counters (although most seem to do it just because get_stats64 is not allowed to sleep, and some hardware requires it to read the counters - this is probably also going to be the case for RTL931x which uses a table instead of direct register access).

However, considering the possible number of ports and the weak CPU on these switches, I'm also not convinced that polling is necessarily a good idea here (but still want to try it).

plappermaul commented:
+1 to avoid polling. Mellanox once had a advisory for MIPS managed infiniband switches (MSX6036) to increase internal polling intervals (not only MIB) because switch became close to unusable with many in use ports.

janh committed:
realtek: implement polling for hardware counters

Maintain 64 bit counters by polling the hardware counters and adding up
the differences. Polling needs to happen just often enough to catch
every single overflow.

As we now have non-overflowing counters now, we can safely calculate
composite counters without getting weird results on overflow. Use this
to follow RFC 3635 more accurately by mapping the hardware counters to
the proper counters, while taking into account hardware quirks as best
as possible.

Signed-off-by: Jan Hoffmann <jan@3e8.eu>
Link: https://github.com/openwrt/openwrt/pull/18415
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

Ansuel commented:
@janh my personal opinion is lets stick with the simple version. If we really want it we should make the poll really long like every 30 second.

Neustradamus commented:
@janh: Any progress on it?

janh commented:
I finally touched this up, so it should be ready for review now.

Changes compared to the previous revision:
- protect reading of 64-bit counters against overflow due to non-atomic read (thanks to @plappermaul for the suggestion)
- increase polling interval to 60 seconds
- add caching of recently read values
- properly handle non-supported devices (RTL931x), so get_stats64 at least returns software counters
- add port range check to all methods
- consider pause frames for calculation of rx/tx_bytes in get_stats64

I also restructured the commits a bit: The first three commits include what is possible to implement without polling. The last commit adds polling to maintain 64-bit counters, also allowing to make some of the reported counters more accurate (as the calculation only makes sense if no overflow can happen).

plappermaul commented:
Hi Jan. 

thanks for the update. This features is greatly apreciated. While this might be annoying I ask you to prefix the touched/new functions and variables to rtldsa_, rtldsa_838x_ ... While pairing this with the old function might look strange it helps to build up a consistent naming schema. Especially getting rid of "rtl83xx" as the previos generic prefix makes clear to what hardware a function belongs.

Recent exmples: https://github.com/openwrt/openwrt/commit/30fae30e038ba6a5fe8184fa339355cfcfd15c43 or https://github.com/openwrt/openwrt/commit/6432b411805fad2e35ad2f3726bcb9d4cd58ce89

Thanks in advance.

janh commented:
Updated to rename struct/functions as suggested by @plappermaul. Otherwise, no changes compared to the last update.

plappermaul commented:
run tested on Zyxel XGS1210-10. Approved, approved & approved 👍 

@robimarko with yesterdays merges there are 3 offsets now.
